### PR TITLE
Fixes #733: Support analysis with only free analyzers

### LINKED ISSUE #733: New Playbook: run only "free" analyzers
*updated: 15th March 2022*

## Idea / Motivation

> An option that allows the user to only run free analyzers, the ones that don't require API Keys or login/password.

The main use-case for this is the following example:
- A user has configured some paid analyzers (with keys/credentials) but maybe for some analysis they want to save on API/service credits by checking the "run only free analyzers" option.

## Pre-requisite(s)

- [ ] #628

## Implementation Idea

This can be solved by creating a playbook that contains all the free (without api key/login) analyzers.

### PR DESCRIPTION
**Pull request by devmrfitz**
# Description

Support for running free analyzers

## Related issues
#733 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [X] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [X] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [ ] The pull request is for the branch `develop`
- [ ] A new analyzer or connector was added, in which case:
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the analyzer/connector provides additional optional configuration).
    - [ ] Secrets were added in [env_file_app_template](https://github.com/intelowlproject/IntelOwl/blob/master/docker/env_file_app_template), [env_file_app_ci](https://github.com/certego/IntelOwl/blob/master/docker/env_file_app_ci) and in the [Installation](./Installation.md) docs, if necessary.
    - [ ] If the analyzer/connector requires mocked testing, `_monkeypatch()` was used in it's class to apply the necessary decorators.
    - [ ] If a File analyzer was added, it's name was explicitly defined in [test_file_scripts.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_file_scripts.py) (not required for Observable Analyzers).
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [X] The tests gave 0 errors.
- [X] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] The commits were squashed into a single one (optional, they will be squashed anyway by the maintainer)
  
### Important Rules
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review

# Real World Example

Please delete if the PR is for bug fixing.
Otherwise, please provide the resulting raw JSON of a finished analysis (and, if you like, a screenshot of the results). This is to allow the maintainers to understand how the analyzer works.

### CODE REVIEW

**Comment by devmrfitz** [comment]
image
image
image
NOTE: I had manually edited the analyzer_config of ClassicDNS to add a secret, and hence make it non-free, for testing purpose

**Comment by devmrfitz** [comment]
@mlodic Please review

**Comment by devmrfitz** [comment]
@mlodic Done

File: api_app/analyzers_manager/controller.py [hunk 1]
```diff
@@ -93,6 +94,11 @@ def filter_analyzers(serialized_data: Dict, warnings: List) -> List[str]:
                 raise NotRunnableAnalyzer(
                     f"{a_name} won't be run because you filtered external analyzers."
                 )
+
+            if free_analyzers_only and len(config.secrets) > 0:
```
Comment (mlodic) [inline api_app/analyzers_manager/controller.py:98]: the check on the `config.secrets` is smart

**Review by mlodic — COMMENTED** [review]
everything makes sense to me but we are missing:

* documentation regarding this feature. Maybe in `Usage.md`

**Comment by mlodic** [comment]
please also change the line `if: contains(github.base_ref, 'develop')` in the PR automation. Sorry but that is the first PR against the new branch

**Comment by devmrfitz** [comment]
@mlodic Review now

**Comment by mlodic** [comment]
Seems good, just one last change I'll ask. Can you move the checkbox you added below the `Select Analyzers` input field. That would make more sense logically from an end user. (Remove "Miscellaneous" also).

Plus I am not sure that "Run only free analyzers" is completely clear as a statement.  I mean, we know that we are filtering out analyzers without secrets but the user could intend this in another way: for instance, if he has added an API key to a service that he got freely, it is normal that he could think that this option won't filter out that analyzer.

Maybe "Run only internal analyzers"? Plus we could add a little text like this one 
image
below that statement which explains when it would make sense to use this option. 
Something like: "use this option to save quotas of configured external services"

(this change would also mean to change pyintelowl (I have already merged that PR so we would need a new one

**Comment by mlodic** [comment]
@devmrfitz after having a discussion with Eshaan, we thought that we could avoid to add this feature cause it will be solved later with the Playbooks. Just adding this right now and remove it later would create too much confusion.

We thank you for your work, was definitevely good.

**Pull request closed without merging**